fhd_core/HEALPix/healpix_snapshot_cube_generate.pro

```diff
@@ -143,86 +143,62 @@ PRO healpix_snapshot_cube_generate,obs_in,status_str,psf_in,cal,params,vis_arr,v
-        FOR pol_i=0,n_pol-1 DO FOR freq_i=0,n_freq_use-1 DO BEGIN
-          *weights_hpx_arr[pol_i,freq_i]=healpix_cnv_apply(Temporary(*weights_arr1[pol_i,freq_i]),hpx_cnv)
-          *variance_hpx_arr[pol_i,freq_i]=healpix_cnv_apply(Temporary(*variance_arr1[pol_i,freq_i]),hpx_cnv)
-;          IF dirty_flag THEN *residual_arr1[pol_i,freq_i]=*dirty_arr1[pol_i,freq_i]-*model_arr1[pol_i,freq_i]
-          IF residual_flag THEN *residual_hpx_arr[pol_i,freq_i]=healpix_cnv_apply(Temporary(*residual_arr1[pol_i,freq_i]),hpx_cnv)
-          IF dirty_flag THEN *dirty_hpx_arr[pol_i,freq_i]=healpix_cnv_apply(Temporary(*dirty_arr1[pol_i,freq_i]),hpx_cnv)
-          IF model_flag THEN *model_hpx_arr[pol_i,freq_i]=healpix_cnv_apply(Temporary(*model_arr1[pol_i,freq_i]),hpx_cnv)
-          *beam_hpx_arr[pol_i,freq_i]=healpix_cnv_apply((*beam_arr[pol_i,freq_i])*nf_vis_use[freq_i],hpx_cnv)
+    beam_squared_cube=fltarr(n_hpx,n_freq_use)
+    weights_cube=fltarr(n_hpx,n_freq_use)
+    variance_cube=fltarr(n_hpx,n_freq_use)
+    IF residual_flag THEN res_cube=fltarr(n_hpx,n_freq_use)
+    IF dirty_flag THEN dirty_cube=fltarr(n_hpx,n_freq_use)
+    IF model_flag THEN model_cube=fltarr(n_hpx,n_freq_use)
+
+    FOR pol_i=0,n_pol-1 DO BEGIN
+
+        FOR freq_i=Long64(0),n_freq_use-1 DO BEGIN
+
+            beam_squared_cube[n_hpx*freq_i]=healpix_cnv_apply((*beam_arr[pol_i,freq_i])*nf_vis_use[freq_i],hpx_cnv)
+            weights_cube[n_hpx*freq_i]=healpix_cnv_apply((*weights_arr1[pol_i,freq_i]),hpx_cnv)
+            variance_cube[n_hpx*freq_i]=healpix_cnv_apply((*variance_arr1[pol_i,freq_i]),hpx_cnv)
+
+            IF residual_flag THEN BEGIN
+                res_cube[n_hpx*freq_i]=healpix_cnv_apply((*residual_arr1[pol_i,freq_i]),hpx_cnv)
+            ENDIF
+            IF dirty_flag THEN BEGIN
+                dirty_cube[n_hpx*freq_i]=healpix_cnv_apply((*dirty_arr1[pol_i,freq_i]),hpx_cnv)
+            ENDIF
+            IF model_flag THEN BEGIN
+                model_cube[n_hpx*freq_i]=healpix_cnv_apply((*model_arr1[pol_i,freq_i]),hpx_cnv)
+            ENDIF
+
         ENDFOR
-    ENDELSE   
-    undefine_fhd,weights_arr1,variance_arr1,residual_arr1,dirty_arr1,model_arr1 ;free memory for beam_arr later!
-    IF iter EQ n_iter-1 THEN undefine_fhd,beam_arr
-    
-    FOR pol_i=0,n_pol-1 DO BEGIN      
-        IF dirty_flag THEN BEGIN
-          dirty_cube=fltarr(n_hpx,n_freq_use)
-            ;write index in much more efficient memory access order
-          FOR fi=Long64(0),n_freq_use-1 DO dirty_cube[n_hpx*fi]=Temporary(*dirty_hpx_arr[pol_i,fi])
-        ENDIF
         
-        IF model_flag THEN BEGIN
-          model_cube=fltarr(n_hpx,n_freq_use)
-          FOR fi=Long64(0),n_freq_use-1 DO model_cube[n_hpx*fi]=Temporary(*model_hpx_arr[pol_i,fi])
+        IF Keyword_Set(save_imagecube) THEN BEGIN
+            save, filename = imagecube_filepath[iter], dirty_arr1, residual_arr1, model_arr1, weights_arr1, variance_arr1, beam_arr, nf_vis_use, obs_out, /compress
```

Comment:
This needs to get moved outside of the polarization `for` loop, or else the same files will be written multiple times.

Reply:
Good catch